packages/features/ee/teams/services/TeamCreationService.ts

```diff
@@ -0,0 +1,379 @@
+import { CreditService } from "@calcom/features/ee/billing/credit-service";
+import stripe from "@calcom/features/ee/payments/server/stripe";
+import { TeamRepository } from "@calcom/features/ee/teams/repositories/TeamRepository";
+import { PermissionCheckService } from "@calcom/features/pbac/services/permission-check.service";
+import { UserRepository } from "@calcom/features/users/repositories/UserRepository";
+import logger from "@calcom/lib/logger";
+import { safeStringify } from "@calcom/lib/safeStringify";
+import slugify from "@calcom/lib/slugify";
+import type { Prisma } from "@calcom/prisma/client";
+import type { CreationSource } from "@calcom/prisma/enums";
+import { MembershipRole } from "@calcom/prisma/enums";
+import { teamMetadataSchema, teamMetadataStrictSchema } from "@calcom/prisma/zod-utils";
+import { inviteMembersWithNoInviterPermissionCheck } from "@calcom/trpc/server/routers/viewer/teams/inviteMember/inviteMember.handler";
+
+import { TRPCError } from "@trpc/server";
+
+const log = logger.getSubLogger({ prefix: ["TeamCreationService"] });
+
+export type CreateTeamsInput = {
+  orgId: number;
+  teamNames: string[];
+  moveTeams: Array<{
+    id: number;
+    newSlug: string | null;
+    shouldMove: boolean;
+  }>;
+  creationSource: CreationSource;
+  ownerId: number;
+};
+
+export type CreateTeamsResult = {
+  duplicatedSlugs: string[];
+};
+
+export type MoveTeamInput = {
+  teamId: number;
+  newSlug?: string | null;
+  org: {
+    id: number;
+    slug: string | null;
+    ownerId: number;
+    metadata: Prisma.JsonValue;
+  };
+  creationSource: CreationSource;
+};
+
+export class TeamCreationService {
+  constructor(
+    private teamRepository: TeamRepository,
+    private creditService: CreditService,
+    private permissionCheckService: PermissionCheckService,
+    private userRepository: UserRepository
+  ) {}
+
+  async createTeamsForOrganization(input: CreateTeamsInput): Promise<CreateTeamsResult> {
+    const { orgId, teamNames, moveTeams, creationSource, ownerId } = input;
+
+    const filteredTeamNames = teamNames.filter((name) => name.trim().length > 0);
+    const organization = await this.validateOrganization(orgId);
+    const duplicatedSlugs = await this.validateTeamSlugs({
+      orgId,
+      teamNames: filteredTeamNames,
+    });
+
+    for (const team of moveTeams.filter((team) => team.shouldMove)) {
+      await this.moveTeamToOrganization({
```

Comment:
what can we do if one of those throw in the middle of moving? (maybe out of the scope of this PR, but just curious to hear your thought)